File: trunk/theLink/example/php/Filter6.php
<?php
#+
#§  \file       theLink/example/php/Filter6.php
#§  \brief      \$Id$
#§  
#§  (C) 2011 - NHI - #1 - Project - Group
#§  
#§  \version    \$Rev$
#§  \author     EMail: aotto1968 at users.berlios.de
#§  \attention  this software has GPL permissions to copy
#§              please contact AUTHORS for additional information
#§

#$stderr = fopen("php://stderr", "w");

class Filter6 extends MqS implements iServerSetup, iServerCleanup, iEvent {
  public function __construct($tmpl=NULL) {
    parent::__construct($tmpl);
    $this->ConfigSetIgnoreExit(TRUE);
    $this->fh = NULL;
  }
  public function ServerSetup() {
    $ftr = $this->ServiceGetFilter();
    $this->ServiceCreate("LOGF", array(&$this, 'LOGF'));
    $this->ServiceCreate("EXIT", array(&$this, 'EXITX'));
    $this->ServiceCreate("+ALL", array(&$this, 'FilterIn'));
    $this->ServiceStorage("PRNT");
    $ftr->ServiceCreate("WRIT",  array(&$ftr , 'WRIT'));
    $ftr->ServiceProxy("+TRT");
  }
  public function ServerCleanup() {
    if ($this->fh != NULL) {
      fclose($this->fh);
    }
    $this->fh = NULL;
  }
  public function WRIT() {
    $fh = $this->ServiceGetFilter()->fh;
    fwrite($fh, $this->ReadC() . "\n");
    fflush($fh);
    $this->SendRETURN();
  }
  public function ErrorWrite() {
    fwrite($this->fh, "ERROR: " . $this->ErrorGetText() . "\n");
    fflush($this->fh);
    $this->ErrorReset();
  }
  public function LOGF() {
    $ftr = $this->ServiceGetFilter();
    if ($ftr->LinkGetTargetIdent() == "transFilter") {
      $this->ReadForward($ftr);
    } else {
      $this->fh = fopen($this->ReadC(), "a");
    }
    $this->SendRETURN();
  }
  public function EXITX() {
    exit(1);
  }
  public function Event() {
    if ($this->StorageCount() == 0) {
      $this->ErrorSetCONTINUE();
    } else {
      $id = 0;
      try {
	$ftr = $this->ServiceGetFilter();
        $ftr->LinkConnect();
	$id = $this->StorageSelect();
	$this->ReadForward($ftr);
      } catch (Exception $ex) {
        $this->ErrorSet($ex);
        if ($this->ErrorIsEXIT()) {
          $this->ErrorReset();
          return;
        } else {
          $this->ErrorWrite();
	}
      }
      $this->StorageDelete($id);
    }
  }
  public function FilterIn() {
    $this->StorageInsert();
    $this->SendRETURN();
  }
}
$srv = FactoryAdd('transFilter', 'Filter6')->New();
try {
  $srv->LinkCreate($argv);
  $srv->ProcessEvent(MqS::WAIT_FOREVER);
} catch (Exception $ex) {
  $srv->ErrorSet($ex);
}
$srv->Exit();

?>

File: trunk/theLink/example/ruby/Filter6.rb
#+
#:  \file       theLink/example/ruby/Filter6.rb
#:  \brief      \$Id$
#:  
#:  (C) 2011 - NHI - #1 - Project - Group
#:  
#:  \version    \$Rev$
#:  \author     EMail: aotto1968 at users.berlios.de
#:  \attention  this software has GPL permissions to copy
#:              please contact AUTHORS for additional information
#:

require "rubymsgque"

class Filter6 < MqS
  attr_accessor :fh
  def initialize(tmpl=nil)
    super(tmpl)
    ConfigSetIgnoreExit(true)
    ConfigSetServerSetup(method(:ServerSetup))
    ConfigSetServerCleanup(method(:ServerCleanup))
    ConfigSetEvent(method(:Event))
    @fh = nil
  end
  def ServerCleanup
    @fh.close if @fh != nil
    @fh = nil
  end
  def ServerSetup
    ftr = ServiceGetFilter()
    ServiceCreate("LOGF", method(:LOGF))
    ServiceCreate("EXIT", method(:EXIT))
    ServiceCreate("+ALL", method(:FilterIn))
    ServiceStorage("PRNT")
    ftr.ServiceCreate("WRIT", ftr.method(:WRIT))
    ftr.ServiceProxy("+TRT")
  end
  def WRIT
    fh = ServiceGetFilter().fh
    fh.write(ReadC() + "\n")
    fh.flush()
    SendRETURN()
  end
  def ErrorWrite
    @fh.write("ERROR: " + ErrorGetText() + "\n")
    @fh.flush
    ErrorReset()
  end
  def LOGF
    ftr = ServiceGetFilter()
    if ftr.LinkGetTargetIdent() == "transFilter" then
      ReadForward(ftr)
    else
      @fh = File.new(ReadC(), "a")
    end
    SendRETURN()
  end
  def EXIT
    Process.exit!(1)
  end
  def Event
    if StorageCount() == 0 then
      ErrorSetCONTINUE()
    else
      id = 0
      begin
	ftr = ServiceGetFilter()
        ftr.LinkConnect
	id = StorageSelect()
	ReadForward(ftr)
      rescue Exception => ex
        ErrorSet(ex)
        if ErrorIsEXIT() then
          ErrorReset()
          return
        else
          ErrorWrite()
	end
      end
      StorageDelete(id)
    end
  end
  def FilterIn
    StorageInsert()
    SendRETURN()
  end
end
srv = FactoryAdd("transFilter", Filter6).New()
begin
  srv.LinkCreate($0,ARGV)
  srv.ProcessEvent(MqS::WAIT_FOREVER)
rescue SignalException => ex
  # ignore
rescue Exception => ex
  srv.ErrorSet(ex)
ensure
  srv.Exit()
end

